docs/src/dev/principles/cpu-abstraction.md

```diff
@@ -0,0 +1,238 @@
+# CPU Abstractions
+
+Core modules in Patina often require divergent implementations for CPU architecture specific mechanisms. This may
+be because the implementation requires assembly which cannot be compiled against other architectures, the implementation
+is only applicable to a given architecture, or both. In these cases, developers are encouraged to ensure that as much
+common code exists as possible. At the point of architecture abstraction, [trait abstraction principles](abstractions.md)
+should be used to define the common contract with the architecture specific logic. This same abstraction can then
+be used for testing of the common logic above/below the abstraction. While there is no one-size-fits-all structure to
+how this should be done, this page describes some simple and common models that should be applied to handling such
+abstractions.
+
+## Code Structure
+
+Implementations should avoid inlining architecture specific code into common logic as this creates code that is difficult
+to read, difficult to test, and is prone to inconsistent contracts between architectures. Architecture specific logic
+should be contained to explicit modules, such as `x64` or `aarch64`. The Patina crate provides macros that can be used
+to add conditional compilation in a consistent way. If the logic is not compilable on other architectures, the module
+should be conditionally compiled for that architecture.
+
+```rust
+#[cfg(target_arch = "x86_64")]
+mod x64 {
+    // A x64 specific module, intended only for platform builds.
+}
+
+#[cfg(target_arch = "aarch64")]
+mod aarch64 {
+    // A x64 specific module, intended only for platform builds.
+}
+```
+
+It should be noted that using just the architecture does cause some inconsistencies around compilation. In particular
+this will allow test compilation, but these tests may not actually be compiled/run since this requires that the
+architecture matches the host system.
+
+### Testing & Documenting
+
+In many cases, the code in the architecture abstractions doesn't contain meaningful logic and merely wraps assembly to
+invoke some CPU functionality not available through rust code. In these cases, there is little to no value in compiling
+for tests as the contained code cannot be tested in usermode or cross architecture or for documentation as the abstraction
+trait should be documented if applicable.
+
+However, when exposing more robust logic that is not entirely architecture abstraction but functionality it is
+preferable that this logic be compiled under test and documentation. An example of this could be the architecture
+specific devices, which are only applicable to a given architecture but should be documented and has complex logic that
+should be tested regardless of the architecture of the test compilation. The `test` and `doc` configuration can be
+added alongside the `aarch64` target architecture. When using this style, it will be necessary for architecture logic to

```

Comment:
```suggestion
added alongside the target architecture. When using this style, it will be necessary for architecture logic to
```
A little confusing to say aarch64 for the generic statement, the example is below to show that